From 66a8a9ed1bb96fdeb8288bb9537b7e1a9c1695c8 Mon Sep 17 00:00:00 2001
From: Michael Braun <michael-dev@fami-braun.de>
Date: Tue, 08 Mar 2011 21:52:56 +0000
Subject: greeter: disable the login button in MODE_AUTHENTICATION if no query is pending

If the user does not need a password to login with gdm, then hitting the
"log in" button causes gdm to crash. This seems to be due to an
AnswerQuery DBUS message being generated even if the password input
field is not visible and an password has not been requested. The user
has a real opportunity to hit this bug at least if pam_mkhomedir is in
use and needs some time to finish due to the size of the skeleton.

This commit desensitizes the Login button when it's unavailable.

https://bugzilla.gnome.org/show_bug.cgi?id=643176
---
Index: gdm-2.32.0/gui/simple-greeter/gdm-greeter-login-window.c
===================================================================
--- gdm-2.32.0.orig/gui/simple-greeter/gdm-greeter-login-window.c	2010-09-15 20:58:52.000000000 +0200
+++ gdm-2.32.0/gui/simple-greeter/gdm-greeter-login-window.c	2011-03-09 19:40:30.803650892 +0100
@@ -458,12 +458,15 @@
         case MODE_SELECTION:
                 /* should never have anything to return to from here */
                 show = FALSE;
+                set_sensitive (login_window, TRUE);
                 break;
         case MODE_TIMED_LOGIN:
                 /* should always have something to return to from here */
                 show = TRUE;
+                set_sensitive (login_window, TRUE);
                 break;
         case MODE_AUTHENTICATION:
+                set_sensitive (login_window, FALSE);
                 if (login_window->priv->num_queries > 1) {
                         /* if we are inside a pam conversation past
                            the first step */
@@ -677,7 +680,6 @@
                 switch_mode (login_window, dialog_mode);
         }
 
-        set_sensitive (login_window, TRUE);
         set_ready (login_window);
         set_focus (GDM_GREETER_LOGIN_WINDOW (login_window));
         update_banner_message (login_window);
